src/prover_disk.hpp

```diff
@@ -238,10 +253,78 @@ class DiskProver {
-    uint8_t k;
+    uint8_t k{0};
     std::vector<uint64_t> table_begin_pointers;
     std::vector<uint64_t> C2;
 
+    void Read(std::vector<uint8_t>& id_in, std::vector<uint8_t>& memo_in, uint8_t& k_in, std::vector<uint64_t>& table_begin_pointers_in, std::vector<uint64_t>& C2_in) const
+    {
+        std::lock_guard<std::mutex> lock(_mtx);
+
+        std::ifstream disk_file(filename, std::ios::in | std::ios::binary);
+
+        if (!disk_file.is_open()) {
+            throw std::invalid_argument("Invalid file " + filename);
+        }
+        struct plot_header header{};
+        // 19 bytes  - "Proof of Space Plot" (utf-8)
+        // 32 bytes  - unique plot id
+        // 1 byte    - k
+        // 2 bytes   - format description length
+        // x bytes   - format description
+        // 2 bytes   - memo length
+        // x bytes   - memo
+
+        SafeRead(disk_file, (uint8_t*)&header, sizeof(header));
+        if (memcmp(header.magic, "Proof of Space Plot", sizeof(header.magic)) != 0)
+            throw std::invalid_argument("Invalid plot header magic");
+
+        uint16_t fmt_desc_len = Util::TwoBytesToInt(header.fmt_desc_len);
+
+        if (fmt_desc_len == kFormatDescription.size() &&
+        !memcmp(header.fmt_desc, kFormatDescription.c_str(), fmt_desc_len)) {
+            // OK
+        } else {
+            throw std::invalid_argument("Invalid plot file format");
+        }
+        id_in = std::vector<uint8_t>(kIdLen);
+        memcpy(id_in.data(), header.id, sizeof(header.id));
+        k_in = header.k;
+        SafeSeek(disk_file, offsetof(struct plot_header, fmt_desc) + fmt_desc_len);
+
+        uint8_t size_buf[2];
+        SafeRead(disk_file, size_buf, 2);
+        memo_in.resize(Util::TwoBytesToInt(size_buf));
+        SafeRead(disk_file, memo_in.data(), memo_in.size());
+
+        table_begin_pointers_in = std::vector<uint64_t>(nTableBeginPointerCount, 0);
+        C2_in = std::vector<uint64_t>();
+
+        uint8_t pointer_buf[8];
+        for (size_t i = 1; i < nTableBeginPointerCount; i++) {
+            SafeRead(disk_file, pointer_buf, 8);
+            table_begin_pointers_in[i] = Util::EightBytesToInt(pointer_buf);
+        }
+
+        SafeSeek(disk_file, table_begin_pointers_in[9]);
+
+        uint8_t c2_size = (Util::ByteAlign(k_in) / 8);
+        uint32_t c2_entries = (table_begin_pointers_in[10] - table_begin_pointers_in[9]) / c2_size;
+        if (c2_entries == 0 || c2_entries == 1) {
+            throw std::invalid_argument("Invalid C2 table size");
+        }
+
+        // The list of C2 entries is small enough to keep in memory. When proving, we can
+        // read from disk the C1 and C3 entries.
+        auto* c2_buf = new uint8_t[c2_size];
+        for (uint32_t i = 0; i < c2_entries - 1; i++) {
+            SafeRead(disk_file, c2_buf, c2_size);
+            C2_in.push_back(Bits(c2_buf, c2_size, c2_size * 8).Slice(0, k).GetValue());
+        }
+
+        delete[] c2_buf;
```

Comment:
this leaks if `SafeRead()` throws an exception. This was an issue before your patch, I just want to make a note to remember to fix it later